File: frontend/lib/data/models/asset.dart
class Asset {
  final String id;
  final String tenantId;
  final String name;
  final String type;
  final String? serialNumber;
  final DateTime? purchaseDate;
  final String status;
  final String condition;
  final String? notes;

  Asset({
    required this.id,
    required this.tenantId,
    required this.name,
    required this.type,
    this.serialNumber,
    this.purchaseDate,
    required this.status,
    required this.condition,
    this.notes,
  });

  factory Asset.fromJson(Map<String, dynamic> json) {
    return Asset(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      name: json['name'] as String,
      type: json['type'] as String,
      serialNumber: json['serial_number'] as String?,
      purchaseDate: json['purchase_date'] != null ? DateTime.tryParse(json['purchase_date'] as String) : null,
      status: json['status'] as String? ?? 'available',
      condition: json['condition'] as String? ?? 'new',
      notes: json['notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'tenant_id': tenantId,
        'name': name,
        'type': type,
        'serial_number': serialNumber,
        'purchase_date': purchaseDate?.toIso8601String(),
        'status': status,
        'condition': condition,
        'notes': notes,
      };
}

class AssetAssignment {
  final String id;
  final String tenantId;
  final String assetId;
  final String employeeId;
  final DateTime assignedDate;
  final DateTime? returnedDate;
  final String? conditionOut;
  final String? conditionIn;
  final String? notes;
  
  // Joined fields
  final String? assetName;
  final String? employeeName;

  AssetAssignment({
    required this.id,
    required this.tenantId,
    required this.assetId,
    required this.employeeId,
    required this.assignedDate,
    this.returnedDate,
    this.conditionOut,
    this.conditionIn,
    this.notes,
    this.assetName,
    this.employeeName,
  });

  factory AssetAssignment.fromJson(Map<String, dynamic> json) {
    return AssetAssignment(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      assetId: json['asset_id'] as String,
      employeeId: json['employee_id'] as String,
      assignedDate: DateTime.parse(json['assigned_date'] as String),
      returnedDate: json['returned_date'] != null ? DateTime.tryParse(json['returned_date'] as String) : null,
      conditionOut: json['condition_out'] as String?,
      conditionIn: json['condition_in'] as String?,
      notes: json['notes'] as String?,
      assetName: json['asset_name'] as String?,
      employeeName: json['employee_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'tenant_id': tenantId,
        'asset_id': assetId,
        'employee_id': employeeId,
        'assigned_date': assignedDate.toIso8601String(),
        'returned_date': returnedDate?.toIso8601String(),
        'condition_out': conditionOut,
        'condition_in': conditionIn,
        'notes': notes,
      };
}

class AssetRequest {
  final String id;
  final String tenantId;
  final String employeeId;
  final String requestedType;
  final String? justification;
  final String status;
  final String priority;
  final String? notes;
  final DateTime? createdAt;
  
  // Joined fields
  final String? employeeName;

  AssetRequest({
    required this.id,
    required this.tenantId,
    required this.employeeId,
    required this.requestedType,
    this.justification,
    required this.status,
    required this.priority,
    this.notes,
    this.createdAt,
    this.employeeName,
  });

  factory AssetRequest.fromJson(Map<String, dynamic> json) {
    return AssetRequest(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      employeeId: json['employee_id'] as String,
      requestedType: json['requested_type'] as String,
      justification: json['justification'] as String?,
      status: json['status'] as String? ?? 'pending',
      priority: json['priority'] as String? ?? 'normal',
      notes: json['notes'] as String?,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at'] as String) : null,
      employeeName: json['employee_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'tenant_id': tenantId,
        'employee_id': employeeId,
        'requested_type': requestedType,
        'justification': justification,
        'status': status,
        'priority': priority,
        'notes': notes,
      };
}

File: backend/internal/repository/asset.go
package repository

import (
	"context"
	"time"

	"hr-saas/internal/db"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type Asset struct {
	ID           uuid.UUID  `json:"id"`
	TenantID     uuid.UUID  `json:"tenant_id"`
	Name         string     `json:"name"`
	Type         string     `json:"type"`
	SerialNumber *string    `json:"serial_number"`
	PurchaseDate *time.Time `json:"purchase_date"`
	Status       string     `json:"status"`
	Condition    string     `json:"condition"`
	Notes        *string    `json:"notes"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
}

type AssetAssignment struct {
	ID           uuid.UUID  `json:"id"`
	TenantID     uuid.UUID  `json:"tenant_id"`
	AssetID      uuid.UUID  `json:"asset_id"`
	EmployeeID   uuid.UUID  `json:"employee_id"`
	AssignedDate time.Time  `json:"assigned_date"`
	ReturnedDate *time.Time `json:"returned_date"`
	ConditionOut *string    `json:"condition_out"`
	ConditionIn  *string    `json:"condition_in"`
	Notes        *string    `json:"notes"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
	// Additional joins for frontend display
	AssetName    string `json:"asset_name,omitempty"`
	EmployeeName string `json:"employee_name,omitempty"`
}

type AssetRequest struct {
	ID            uuid.UUID `json:"id"`
	TenantID      uuid.UUID `json:"tenant_id"`
	EmployeeID    uuid.UUID `json:"employee_id"`
	RequestedType string    `json:"requested_type"`
	Justification *string   `json:"justification"`
	Status        string    `json:"status"`
	Priority      string    `json:"priority"`
	Notes         *string   `json:"notes"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
	// Additional joins for frontend display
	EmployeeName string `json:"employee_name,omitempty"`
}

// ---------------------------------------------------------
// ASSETS
// ---------------------------------------------------------

func ListAssets(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]Asset, error) {
	var list []Asset
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT id, tenant_id, name, type, serial_number, purchase_date, status, condition, notes, created_at, updated_at
			 FROM assets WHERE deleted_at IS NULL ORDER BY name ASC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var a Asset
			if err := rows.Scan(&a.ID, &a.TenantID, &a.Name, &a.Type, &a.SerialNumber, &a.PurchaseDate, &a.Status, &a.Condition, &a.Notes, &a.CreatedAt, &a.UpdatedAt); err != nil {
				return err
			}
			list = append(list, a)
		}
		return rows.Err()
	})
	if err != nil {
		return nil, err
	}
	return list, nil
}

func CreateAsset(ctx context.Context, pool *pgxpool.Pool, tenantID uuid.UUID, asset *Asset) (*Asset, error) {
	var created Asset
	err := db.WithTenant(ctx, pool, tenantID.String(), func(ctx context.Context, tx pgx.Tx) error {
		return tx.QueryRow(ctx,
			`INSERT INTO assets (tenant_id, name, type, serial_number, purchase_date, status, condition, notes)
			 VALUES ($1, $2, $3, $4, $5, COALESCE($6, 'available'), COALESCE($7, 'new'), $8)
			 RETURNING id, tenant_id, name, type, serial_number, purchase_date, status, condition, notes, created_at, updated_at`,
			tenantID, asset.Name, asset.Type, asset.SerialNumber, asset.PurchaseDate, asset.Status, asset.Condition, asset.Notes,
		).Scan(&created.ID, &created.TenantID, &created.Name, &created.Type, &created.SerialNumber, &created.PurchaseDate, &created.Status, &created.Condition, &created.Notes, &created.CreatedAt, &created.UpdatedAt)
	})
	if err != nil {
		return nil, err
	}
	return &created, nil
}

func UpdateAsset(ctx context.Context, pool *pgxpool.Pool, tenantID string, id string, asset *Asset) (*Asset, error) {
	var updated Asset
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		return tx.QueryRow(ctx,
			`UPDATE assets
			 SET name = $1, type = $2, serial_number = $3, purchase_date = $4, status = $5, condition = $6, notes = $7, updated_at = now()
			 WHERE id = $8 AND deleted_at IS NULL
			 RETURNING id, tenant_id, name, type, serial_number, purchase_date, status, condition, notes, created_at, updated_at`,
			asset.Name, asset.Type, asset.SerialNumber, asset.PurchaseDate, asset.Status, asset.Condition, asset.Notes, id,
		).Scan(&updated.ID, &updated.TenantID, &updated.Name, &updated.Type, &updated.SerialNumber, &updated.PurchaseDate, &updated.Status, &updated.Condition, &updated.Notes, &updated.CreatedAt, &updated.UpdatedAt)
	})
	if err != nil {
		return nil, err
	}
	return &updated, nil
}

func DeleteAsset(ctx context.Context, pool *pgxpool.Pool, tenantID string, id string) error {
	return db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		_, err := tx.Exec(ctx, `UPDATE assets SET deleted_at = now() WHERE id = $1`, id)
		return err
	})
}

// ---------------------------------------------------------
// ASSET ASSIGNMENTS
// ---------------------------------------------------------

func AssignAsset(ctx context.Context, pool *pgxpool.Pool, tenantID uuid.UUID, assignment *AssetAssignment) (*AssetAssignment, error) {
	var created AssetAssignment
	err := db.WithTenant(ctx, pool, tenantID.String(), func(ctx context.Context, tx pgx.Tx) error {
		// 1. Create the assignment
		err := tx.QueryRow(ctx,
			`INSERT INTO asset_assignments (tenant_id, asset_id, employee_id, assigned_date, condition_out, notes)
			 VALUES ($1, $2, $3, COALESCE($4, CURRENT_DATE), $5, $6)
			 RETURNING id, tenant_id, asset_id, employee_id, assigned_date, returned_date, condition_out, condition_in, notes, created_at, updated_at`,
			tenantID, assignment.AssetID, assignment.EmployeeID, assignment.AssignedDate, assignment.ConditionOut, assignment.Notes,
		).Scan(&created.ID, &created.TenantID, &created.AssetID, &created.EmployeeID, &created.AssignedDate, &created.ReturnedDate, &created.ConditionOut, &created.ConditionIn, &created.Notes, &created.CreatedAt, &created.UpdatedAt)
		if err != nil {
			return err
		}

		// 2. Update asset status to 'assigned'
		_, err = tx.Exec(ctx, `UPDATE assets SET status = 'assigned', updated_at = now() WHERE id = $1`, assignment.AssetID)
		return err
	})
	if err != nil {
		return nil, err
	}
	return &created, nil
}

func ReturnAsset(ctx context.Context, pool *pgxpool.Pool, tenantID string, assignmentID string, conditionIn string) error {
	return db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		var assetID uuid.UUID
		err := tx.QueryRow(ctx,
			`UPDATE asset_assignments
			 SET returned_date = CURRENT_DATE, condition_in = $1, updated_at = now()
			 WHERE id = $2
			 RETURNING asset_id`,
			conditionIn, assignmentID,
		).Scan(&assetID)
		if err != nil {
			return err
		}

		// Update asset status to 'available' and set new condition
		_, err = tx.Exec(ctx, `UPDATE assets SET status = 'available', condition = $1, updated_at = now() WHERE id = $2`, conditionIn, assetID)
		return err
	})
}

func ListAssetAssignments(ctx context.Context, pool *pgxpool.Pool, tenantID string, employeeID *string) ([]AssetAssignment, error) {
	var list []AssetAssignment
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		query := `
			SELECT aa.id, aa.tenant_id, aa.asset_id, aa.employee_id, aa.assigned_date, aa.returned_date, aa.condition_out, aa.condition_in, aa.notes, aa.created_at, aa.updated_at,
			       a.name as asset_name, e.name as employee_name
			FROM asset_assignments aa
			JOIN assets a ON aa.asset_id = a.id
			JOIN employees e ON aa.employee_id = e.id
			WHERE 1=1
		`
		var args []interface{}
		var argCount int = 1

		if employeeID != nil && *employeeID != "" {
			query += ` AND aa.employee_id = $` + javaToPgPlaceholder(argCount)
			args = append(args, *employeeID)
			argCount++
		}

		query += ` ORDER BY aa.assigned_date DESC`

		rows, err := tx.Query(ctx, query, args...)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var a AssetAssignment
			if err := rows.Scan(&a.ID, &a.TenantID, &a.AssetID, &a.EmployeeID, &a.AssignedDate, &a.ReturnedDate, &a.ConditionOut, &a.ConditionIn, &a.Notes, &a.CreatedAt, &a.UpdatedAt, &a.AssetName, &a.EmployeeName); err != nil {
				return err
			}
			list = append(list, a)
		}
		return rows.Err()
	})
	if err != nil {
		return nil, err
	}
	return list, nil
}

// ---------------------------------------------------------
// ASSET REQUESTS
// ---------------------------------------------------------

func CreateAssetRequest(ctx context.Context, pool *pgxpool.Pool, tenantID uuid.UUID, req *AssetRequest) (*AssetRequest, error) {
	var created AssetRequest
	err := db.WithTenant(ctx, pool, tenantID.String(), func(ctx context.Context, tx pgx.Tx) error {
		return tx.QueryRow(ctx,
			`INSERT INTO asset_requests (tenant_id, employee_id, requested_type, justification, status, priority, notes)
			 VALUES ($1, $2, $3, $4, COALESCE($5, 'pending'), COALESCE($6, 'normal'), $7)
			 RETURNING id, tenant_id, employee_id, requested_type, justification, status, priority, notes, created_at, updated_at`,
			tenantID, req.EmployeeID, req.RequestedType, req.Justification, req.Status, req.Priority, req.Notes,
		).Scan(&created.ID, &created.TenantID, &created.EmployeeID, &created.RequestedType, &created.Justification, &created.Status, &created.Priority, &created.Notes, &created.CreatedAt, &created.UpdatedAt)
	})
	if err != nil {
		return nil, err
	}
	return &created, nil
}

func ListAssetRequests(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]AssetRequest, error) {
	var list []AssetRequest
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT ar.id, ar.tenant_id, ar.employee_id, ar.requested_type, ar.justification, ar.status, ar.priority, ar.notes, ar.created_at, ar.updated_at,
			        e.name as employee_name
			 FROM asset_requests ar
			 JOIN employees e ON ar.employee_id = e.id
			 ORDER BY ar.created_at DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r AssetRequest
			if err := rows.Scan(&r.ID, &r.TenantID, &r.EmployeeID, &r.RequestedType, &r.Justification, &r.Status, &r.Priority, &r.Notes, &r.CreatedAt, &r.UpdatedAt, &r.EmployeeName); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	if err != nil {
		return nil, err
	}
	return list, nil
}

func UpdateAssetRequestStatus(ctx context.Context, pool *pgxpool.Pool, tenantID string, requestID string, status string) error {
	return db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		_, err := tx.Exec(ctx, `UPDATE asset_requests SET status = $1, updated_at = now() WHERE id = $2`, status, requestID)
		return err
	})
}

// helper for string conversion
func javaToPgPlaceholder(i int) string {
	importStr := []string{"1", "2", "3", "4", "5", "6", "7", "8", "9"}
	if i > 0 && i <= len(importStr) {
		return importStr[i-1]
	}
	return "1" // fallback
}
